docs/proposals/11-cluster-health.md

```diff
@@ -0,0 +1,205 @@
+
+This is just a sample set of condition types that might be of immediate relevance. It is not an exhaustive list.
+The suggested possible actions for these condition types are also just samples and not an exhaustive list.
+
+#### KubeAPIServerUnreachableExternally
+
+The monitoring (probing) part of the current `probe` [sub-command](https://github.com/gardener/dependency-watchdog/blob/master/cmd/probe.go) of the [dependency-watchdog](ttps://github.com/gardener/dependency-watchdog) can be re-factored into a separate component that updates the conditions with the `KubeAPIServerUnreachableExternally` condition type.
+
+If there is a way to detect the exact reason for this condition (for example, the loadbalancer is down), then that information could be published as a separate condition type.
+
+##### Possible Actions
+
+- Alerts can be raised to the administrator of the cluster as well as the operator/support team on duty if this condition lasts for some duration.
+- The `kube-controller-manager` could be scaled down to avoid marking `nodes` and `pods` as `NotReady` unnecessarily.
+This could be done by the scaler part of the current `probe` [sub-command](https://github.com/gardener/dependency-watchdog/blob/master/cmd/probe.go) of the [dependency-watchdog](https://github.com/gardener/dependency-watchdog) but re-factored as a separate scaler component which scales some targets based on some configured conditions in the `Cluster` resource. 
+
+#### ETCDUnreachable
+
+The `Endpoint` monitoring part of the `root` [command](https://github.com/gardener/dependency-watchdog/blob/master/cmd/root.go) of the [dependency-watchdog](https://github.com/gardener/dependency-watchdog) can be re-factored into a separate component that updates the conditions with the `ETCDUnreachable` condition type.
+
+##### Possible Actions
+
+- Alerts can be raised to the administrator of the cluster as well as the operator/support team on duty if this condition lasts for some duration.
+- When this condition transitions from status `true` to `false` (i.e. `etcd` transitions from being unreachable to reachable), the dependent `pods` could be deleted if they are in `CrashLoopBackoff`. This is to make sure the dependent `pods` recover as fast as possible from an `etcd` down time. This could be done by the restarter part of the current `root` [command](https://github.com/gardener/dependency-watchdog/blob/master/cmd/root.go) of the [dependency-watchdog](https://github.com/gardener/dependency-watchdog) but re-factored as a separate restarter component which restarts (deletes) some targets `pods` based on some configured conditions in the `Cluster` resource but only if the `pods` are in `CrashLoopBackoff`. 
+
+#### KubeAPIServerUnreachableInternally
+
+The `KubeAPIServerUnreachableInternally` condition type can be implemented (both the monitoring and the remedial action parts) similar to the [`ETCDUnreachable`](#etcdunreachable) condition type.
```

Comment:
This condition is useful for the [scaler](https://github.com/gardener/dependency-watchdog/blob/master/pkg/scaler/prober.go#L215) as a safety check to avoid scaling down the `kube-controller-manager` if both internal and external endpoints to the `kube-apiserver` are down. The scale down is done only if the internal endpoint is healthy and the external endpoint is unhealthy.

Reply:
I.e. some conditions might be of use to avoid taking action.